### ISSUE TITLE
Narrow non-literal type to literal based on equality check

### ISSUE THREAD

**Issue by alexmacnorthstar** [astral-sh/ty#1566]
### Summary

Using a guard to check if a str is a value from a list of literals like this:

```py
from typing import Literal

type MyType = Literal['test1', 'test2']

def test(foo:str) -> MyType:
  valid_vals: list[MyType] = ['test1', 'test2']
  if foo in valid_vals:
    return foo
  return 'test1'
```

Gives:

```
error[invalid-return-type]: Return type does not match returned value
 --> ty_test.py:5:22
  |
3 | type MyType = Literal['test1', 'test2']
4 |
5 | def test(foo:str) -> MyType:
  |                      ------ Expected `MyType` because of return type
6 |   valid_vals: list[MyType] = ['test1', 'test2']
7 |   if foo in valid_vals:
8 |     return foo
  |            ^^^ expected `MyType`, found `str`
9 |   return 'test1'
  |
info: rule `invalid-return-type` is enabled by default

Found 1 diagnostic
```

But foo is constrained to only valid values of the MyType literal at this point

### Version

ty 0.0.1-alpha.26 (b225fd8b4 2025-11-10)

**Comment by carljm** [comment]
It's intentional that we haven't so far implemented support for this kind of narrowing, because it is unsound. The type `str` includes subclasses of `str`, which are permitted to implement `__eq__` to compare equal to whatever they like. And a subclass of `str` that compares equal to `"test1"` does not inhabit the type `Literal["test1"]` -- that type is limited to the instance of the actual `str` class with value `"test1"`.

```py
class MyStr(str):
    def __eq__(self, other):
        return True  # YOLO

test(MyStr("something else"))  # returns MyStr("something else") at runtime, violating the return type of `test`
```

If the type of `foo` were a larger union of literal types, we would eliminate types from that union which we know _cannot_ compare equal to `test1`. We currently do this for `in` checks only if the RHS is a tuple, not if its a list.

Pyright narrows in this case; mypy does not.

I will leave this issue open for now, to consider whether a change in behavior makes sense (and provide a central place for user feedback on the question.)

**Comment by carljm** [comment]
We could add a diagnostic that prohibits overriding `__eq__` on subclasses of all types for which we have literal types, and then go ahead and implement this kind of narrowing. See #2171 where we plan to do this for tuples. I don't know if overriding `__eq__` on `int` or `str` subclasses is more common than on `tuple` subclasses.

The advantage of this approach is that people who aren't worried about the unsoundness and want to match pyright behavior can always just turn off the "unsafe subclass" warning and still get the narrowing.

**Comment by yilei** [comment]
We have a similar use case (several tens of errors in our codebase) for `dict[Literal["a", "b", ...], T]`, something like:

```python
from typing import Literal, TypeAlias

ConfigType: TypeAlias = Literal["dev", "staging", "prod"]  # Reusable alias
CONFIGS: dict[ConfigType, int] = {"dev": 1, "staging": 2, "prod": 3}


def get_config(name: str) -> int | None:
    if name in CONFIGS:
        # error[invalid-argument-type]: Method `__getitem__` of type `bound method dict[Literal["dev", "staging", "prod"], int].__getitem__(key: Literal["dev", "staging", "prod"], /) -> int` cannot be called with key of type `str` on object of type `dict[Literal["dev", "staging", "prod"], int]`
        return CONFIGS[name]
    # Do something else ...
    return None
```

**Comment by carljm** [comment]
@yilei The only thing ty needs in order to be happy with that already is for it to be `name: LiteralString` instead of `name: str`, which excludes badly-behaved `str` subclasses. Don't know if that's feasible for you or not.

**Comment by andreasoprani** [comment]
I think I have a problem that is adjacent to this one with a tuple and literal ints, here's the code:
```py
from typing import Any, Literal


def foo(v: Any) -> Literal[1, 2, 3] | None:
    if not v:
        return None

    if isinstance(v, str):
        v = int(v)

    if v not in (1, 2, 3):
        raise ValueError("v not in (1, 2, 3)")

    return v
```

Pyright has no issues with this code while ty says:
```
error[invalid-return-type]: Return type does not match returned value
  --> literalint.py:14:12
   |
12 |         raise ValueError("v not in (1, 2, 3)")
13 |
14 |     return v
   |            ^ expected `Literal[1, 2, 3] | None`, found `(Any & ~AlwaysFalsy & ~str) | int`
   |
  ::: literalint.py:4:20
   |
 4 | def foo(v: Any) -> Literal[1, 2, 3] | None:
   |                    ----------------------- Expected `Literal[1, 2, 3] | None` because of return type
 5 |     if not v:
 6 |         return None
   |
info: rule `invalid-return-type` is enabled by default
```

Ty version: `0.0.15 (eb43dff06 2026-02-04)`

**Comment by carljm** [comment]
@andreasoprani Yes, this is an interesting case, thanks for reporting it. In this case your code is perfectly safe; ty doesn't consider that since you just called `int(v)` that means the "could be int" portion of the union is definitely exactly `int`; it can't (in this case) be a subclass of `int` with surprising `__eq__` behavior.

**Comment by ultrabear** [comment]
Just ran into another variant of this case:
```py
from typing import Literal

type L = Literal["a", "b"]

S: frozenset[L] = frozenset({"a", "b"})


def chk(s: str) -> L | None:
    if s in S:
        return s
```
ignoring that ty errors on the frozenset construction (separate issue)

Personally, I would be happy with an explicit flag that makes certain python operations UB in exchange for typechecking general code better

**Comment by yilei** [comment]
> [@yilei](https://github.com/yilei) The only thing ty needs in order to be happy with that already is for it to be `name: LiteralString` instead of `name: str`, which excludes badly-behaved `str` subclasses. Don't know if that's feasible for you or not.

Unfortunately it's not possible for most of the cases. `get_config()` could be called with a `name: str` variable coming from e.g. an env var, or some other functions deep in the call graph.

**Comment by maximemulder** [comment]
Encountered this case in my codebase as well, like some others I would be happy with an option to type check it.

In the meantime I am curious, is there a better workaround to type check this kind of code (without restoring to `# ty: ignore`) than exhaustively checking and returning each possible literal ?

```py
def foo(p: str) -> Literal['aaa', 'bbb']:
    if p in ['aaa', 'bbb']:
        return p

    raise Exception()
```

**Comment by sharkdp** [comment]
@maximemulder If `p: LiteralString` is an option for you, you can do
```py
def foo(p: LiteralString) -> Literal['aaa', 'bbb']:
    if p in ('aaa', 'bbb'):
        return p

    raise Exception()
```

**Comment by maximemulder** [comment]
Not applicable in my case @sharkdp but thanks for the solution nonetheless. I guess I'll just resort to exhaustively listing each literal (I don't have too many so it's ok).

**Comment by carljm** [comment]
> We could add a diagnostic that prohibits overriding `__eq__` on subclasses of all types for which we have literal types, and then go ahead and implement this kind of narrowing. See [#2171](https://github.com/astral-sh/ty/issues/2171) where we plan to do this for tuples. I don't know if overriding `__eq__` on `int` or `str` subclasses is more common than on `tuple` subclasses.

I think we should do this. Since there weren't any objections earlier, going to go ahead and remove `needs-decision` tag here.

**Comment by AlexWaygood** [comment]
> > We could add a diagnostic that prohibits overriding `__eq__` on subclasses of all types for which we have literal types, and then go ahead and implement this kind of narrowing. See [#2171](https://github.com/astral-sh/ty/issues/2171) where we plan to do this for tuples. I don't know if overriding `__eq__` on `int` or `str` subclasses is more common than on `tuple` subclasses.
> 
> I think we should do this.

Hmm, I object 😆

This issue has very few upvotes considering how long it's been open. It's true that we've had _lots_ of duplicate issues filed for this, but my experience triaging them has been that nearly all reporters understand our reasoning when we explain it to them and usually agree with it. The main problem here, as I see it, is that ty's behaviour is not what users initially expect it to be, and so they often report issues regarding it -- but that could equally well be tackled by improving our diagnostics rather than changing our behaviour, and we haven't even tried doing improving our diagnostics yet.

> We could add a diagnostic that prohibits overriding `__eq__` on subclasses of all types for which we have literal types

This approach makes equality narrowing sound as long as you only have first-party code in your project. Most projects don't, however: most projects have third-party dependencies, and we won't emit a diagnostic if a third-party dependency unsoundly overrides `__eq__`. This could make equality narrowing in your first-party code which _uses_ that class just as unsound as if you'd overridden `__eq__` in your first-party code.

In some situations we _have_ to resort to a technique like this in order to achieve a modicum of soundness. This is the technique we use for emitting diagnostics on Liskov violations, for example -- there's no other way to attempt soundness there. But I feel like this strategy should be a last resort.

Our type narrowing is already unsound in many ways: just like every other type checker, we narrow on subscript and attribute expressions. We also already do narrowing [here](https://play.ty.dev/f72fec00-1737-43b2-9fc1-7b67fb286c26), even though a tuple subclass could easily override `__contains__`. If we really feel strongly that we should change our behaviour here without even trying to improve our diagnostics first, I think I'd advocate just accepting that type narrowing in Python is inherently unsound, and not bother with the diagnostics for overridden `__eq__` methods (which probably won't meaningfully improve soundness anyway, given the third-party code problem). Attempting to detect overridden `__eq__` methods feels like a halfway house that will still annoy users, still leave us unsound, and still leave us incompatible with other type checkers.

**Comment by carljm** [comment]
> The main problem here, as I see it, is that ty's behaviour is not what users initially expect it to be, and so they often report issues regarding it -- but that could equally well be tackled by improving our diagnostics rather than changing our behaviour, and we haven't even tried doing improving our diagnostics yet.

I don't agree. IMO the main problem here is not one of understanding or diagnostic quality, but the fact that there are useful and reasonable code patterns (examples above) which one might want to write (and empirically, people often do want to write), which depend for soundness only on the knowledge that you aren't passing around badly-behaved `str` subclasses in your codebase (which is likely true for the vast majority of people), which we simply cannot type check, nor make any useful alternative recommendation for how to type check. The best we've been able to suggest is "annotate the incoming value as `LiteralString`", which is usually not feasible either.  

> This approach makes equality narrowing sound as long as you only have first-party code in your project.

You point out that this is also true of Liskov, which is required for the soundness of easily the majority of Python type checker feedback. All protocol assignment, all attribute access, all method calls -- all depend for their soundness on Liskov validity. So I think this characteristic you observe is closer to "the fundamental nature of the entire Python type checking project" than it is "a narrowly applied last resort."

> Our type narrowing is already unsound in many ways: just like every other type checker, we narrow on subscript and attribute expressions.

Yes, which is why it feels strange and inconsistent to me that we would hold the line on soundness here, which I think is no more likely (probably less likely) to cause real-world problems than our existing handling of attribute and subscript narrowing.

> I think I'd advocate just accepting that type narrowing in Python is inherently unsound, and not bother with the diagnostics for overridden `__eq__` methods

That's fine with me, too.

I think that the diagnostics on overridden `__eq__` are useful and worth having in their own right (though arguably closer to "lint rule" than "type error") regardless of whether you think they are adequate to consider this narrowing fully sound. They indicate that you are doing something which is likely to violate the assumptions of every Python type checker, not to mention the assumptions of every reader of the code. That seems worth warning about. Like any other dubious-but-maybe-sometimes-justified practice, you can silence that error code if you don't want it, or `type: ignore` the specific violation. 

But I'd be happy with allowing this narrowing without that lint rule, too.

**Comment by maximemulder** [comment]
> The main problem here, as I see it, is that ty's behaviour is not what users initially expect it to be, and so they often report issues regarding it -- but that could equally well be tackled by improving our diagnostics rather than changing our behaviour, and we haven't even tried doing improving our diagnostics yet.

As a user I agree diagnostics are a problem. I did not know what the problem was and almost opened a duplicate issue until I found this issue.

In my opinion the ideal solution is to have a rule to allow or disallow unsound literal narrowing, have it allowed by default, and output appropriate diagnostics otherwise.

Ideally there would be such such rules/diagnostics for every unsound type checking feature, but I don't know how feasible that is.

I see some value in a warning on overriding `__eq__` for literal subclasses but I don't think it should be tied to the narrowing here. Personally, I only ever encountered such classes in third-party code.

**Comment by AlexWaygood** [comment]
> I don't agree. IMO the main problem here is not one of understanding or diagnostic quality, but the fact that there are useful and reasonable code patterns (examples above) which one might want to write (and empirically, people often do want to write), which depend for soundness only on the knowledge that you aren't passing around badly-behaved `str` subclasses in your codebase (which is likely true for the vast majority of people)

I agree that these code patterns exist, and I agree that we've had lots of people who want to write them. I still think the empirical evidence is slim that this has caused a major problem for our users, though. Again, we've had _lots_ of duplicates of this issue, but _only two_ upvotes on this issue itself despite the fact that it's been open since November. This is very different from some other issues where we've had lots of duplicates reported, some of which have been (or are) very highly upvoted. This suggests to me that most users are able to find easy workarounds once they understand the reason for our behaviour. From what I can see, it's been one of our more successful experiments.

> which we simply cannot type check, nor make any useful alternative recommendation for how to type check. The best we've been able to suggest is "annotate the incoming value as `LiteralString`", which is usually not feasible either.

Though we have floated, a number of times, the idea of tracking internally whether an object is an instance of "exactly `str`" vs "`str` or any subclass". If we did that, we could soundly allow narrowing for this:

```py
def f(x: object):
    if type(x) is str and x == "foo":
        reveal_type(x)  # revealed: Literal["foo"]
```

because the first guard would allow us to narrow to `Exactly[str]` and the second guard would allow us to narrow from that to `Literal["foo"]`.

**Comment by carljm** [comment]
I have a higher bar for a successful experiment than "not a major problem" and "users seem able to find workarounds." I checked all issues that have been closed as duplicates of this, and can't find a single example of anyone expressing positive appreciation for this behavior. In addition to the two upvotes on the issue itself, there are five hearts on my initial comment suggesting we could allow this narrowing.

I think the burden of proof needs to be the other way around here. If we will diverge from the accepted behavior of all other type checkers, we should have evidence that our choice is an improvement for users. "Other things seem to have caused more problems for our users" is tepid endorsement.

Again: what is the rationale for insisting on soundness in equality narrowing, but still performing unsound subscript and attribute narrowing? The workaround in the latter case is even clearer: assign to a local variable.

I would be fine with an option to disable unsound equality narrowing, but I agree with @maximemulder that the default should be to allow it.

I don't think the added complexity and overhead of tracking exclude-subclass types (or making users write `type(x) is str`) is worth it.

**Comment by sharkdp** [comment]
We discussed this internally and concluded that it's not worth trying to keep our current correct-but-overly-pedantic behavior, especially since there is no good alternative solution for users.

So the plan here is to allow unsafe narrowing on equality checks.

For users that want the pedantic behavior, we could consider implementing a new opt-in rule. This would either issue a diagnostic on equality checks that lead to unsafe narrowing, or on `__eq__` overrides in `str/int/bytes` subclasses. Both approaches are very prone to both false positives and false negatives, so it's currently not clear if this is a good idea. I suggest that we open a ticket to address this once the main feature here (the unsafe narrowing) is implemented.

**Comment by alexchandel** [comment]
Maybe there should be a `pedantic` option for this level of strictness, where technically unsound narrowings are disabled?  At the very least, the docs need a page listing which "unsound" type inference rules are applied.  Because it's a learning experience.  The average user who's testing `a_str in some_literals` doesn't realize subclass `__eq__` could subvert the type system.

And I get why it's unsound. What is the canonically "sound" way to narrow a non-literal to a literal, given some valid set of literals?

**Comment by carljm** [comment]
> What is the canonically "sound" way to narrow a non-literal to a literal, given some valid set of literals?

`if type(x) is str and x == "foo"` is sufficient -- but ty currently doesn't narrow for this either.

As mentioned above our plan here is to simply do this narrowing, even though it's technically not sound. We just haven't gotten around to implementing that yet.

**Comment by alexchandel** [comment]
Is there a list of "technically unsound" narrowings that ty does in the docs somewhere?  These would be good to know, as someone who has had to subclass `str` before.

And it seems like astral will be in a position to make an optimized Python interpreter for strictly-typed Python soon.  Will these unsound narrowings affect that?
